File: app/src/main/java/com/udacity/astroapp/utils/DateTimeUtils.java
package com.udacity.astroapp.utils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

public class DateTimeUtils {

    private static final SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH);

    public static String getCurrentLocalDate() {
        return formatter.format(getCalendar().getTime());
    }

    public static String getPreviousDate(String localDate) {
        Calendar calendar = getCalendar();
        try {
            calendar.setTime(Objects.requireNonNull(formatter.parse(localDate)));
            calendar.add(Calendar.DATE, -1);
            System.out.println(formatter.format(calendar.getTime()));
            return formatter.format(calendar.getTime());
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Calendar getCalendar() {
        return Calendar.getInstance();
    }

    public static String getFormattedDate(Date date) {
        return formatter.format(date);
    }

    public static int getYear(String date) {
        return Integer.parseInt(date.substring(0, 4));
    }

    public static int getMonth(String date) {
        String monthString = date.substring(5, 7);
        int month = Integer.parseInt(monthString);
        month = month - 1;
        return month;
    }

    public static int getDay(String date) {
        return Integer.parseInt(date.substring(8, 10));
    }

    public static String getFormattedDateFromString (String date) {
        return date.substring(0, 10);
    }
}

File: app/src/main/kotlin/com/udacity/astroapp/utils/DateTimeUtils.kt
package com.udacity.astroapp.utils

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

object DateTimeUtils {

    private val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)

    fun getCurrentLocalDate(): String {
        return formatter.format(getCalendar().time)
    }

    fun getPreviousDate(localDate: String): String? {
        val calendar = getCalendar()
        return try {
            calendar.time = formatter.parse(localDate) ?: return null
            calendar.add(Calendar.DATE, -1)
            println(formatter.format(calendar.time))
            formatter.format(calendar.time)
        } catch (e: ParseException) {
            e.printStackTrace()
            null
        }
    }

    fun getCalendar(): Calendar {
        return Calendar.getInstance()
    }

    fun getFormattedDate(date: Date): String {
        return formatter.format(date)
    }

    fun getYear(date: String): Int {
        return date.substring(0, 4).toInt()
    }

    fun getMonth(date: String): Int {
        return date.substring(5, 7).toInt()
    }

    fun getDay(date: String): Int {
        return date.substring(8, 10).toInt()
    }

    fun getNextDate(localDate: String): String? {
        val calendar = getCalendar()
        return try {
            calendar.time = formatter.parse(localDate) ?: return null
            calendar.add(Calendar.DATE, 1)
            formatter.format(calendar.time)
        } catch (e: ParseException) {
            e.printStackTrace()
            null
        }
    }

    fun getDateFromCalendar(year: Int, month: Int, dayOfMonth: Int): String {
        val calendar = getCalendar()
        calendar.set(year, month, dayOfMonth)
        return formatter.format(calendar.time)
    }
}